src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,393 @@
+    param code_as_comment: All lines of the generated source code will be prefixed with '#' in which case code
+     are turned in comments. This is the default behavior.
+    """
+
+    try:
+        check_func_type(func)
+    except TypeError:
+        if code_as_comment:
+            # if it is an unsupported type, then no code generation and return emtpy string
+            return ""
+        raise
+
+    # stored referenced object, key is the object name, value is the object
+    ref_objects = {}
+    # stored modules, each item should be a tuple of two strings, first is the true module name, second is the used name
+    # such as alias or just the name
+    to_import = set()
+    # imports class/funcs/vars form other modules, each key is the module name
+    # each item is a set of tuples of two strings as the to_import, first module name, second alias
+    to_import_from_module = defaultdict(set)
+    # classes that should be generated source code
+    classes_to_generate = []
+
+    header_text = CODE_AS_COMMENT_HINT if code_as_comment else CODE_HEADER
+    global_vars_text = ""
+    func_text = ""
+    classes_text = ""
+
+    func_module_name = func.__module__
+
+    # 1. Resolve objects referenced by functions including classes, methods, modules, global variables
+    if isinstance(func, FunctionType):
+        get_func_references(func, ref_objects)
+        to_import = to_import.union(
+            extract_submodule_imports(
+                func, [v for v in ref_objects.values() if isinstance(v, ModuleType)]
+            )
+        )
+        for v in [
+            v
+            for v in ref_objects.values()
+            if inspect.isclass(v) and v.__module__ == func_module_name
+        ]:
+            get_class_references(v, func_module_name, ref_objects, classes_to_generate)
+    elif isinstance(func, BuiltinFunctionType):
+        if func_module_name != builtins.__name__:
+            to_import.add((func_module_name, func_module_name))
+
+    # 2. Dealing with the referenced objects by types
+    for name, obj in ref_objects.items():
+        if obj == func:
+            continue
+        if inspect.ismodule(obj):
+            # a) imported modules
+            to_import.add((obj.__name__, name))  # name could be an alias
+        elif (
+            inspect.isclass(obj) or inspect.isfunction(obj)
+        ) and obj.__module__ != func_module_name:
+            # b) classes or functions imported from other modules
+            to_import_from_module[obj.__module__].add(
+                (obj.__name__, name)
+            )  # name could be an alias
+        else:
+            # function/class/variables defined in the same module
+            if inspect.isfunction(obj):
+                func_text += textwrap.dedent(inspect.getsource(obj))
+            elif inspect.isclass(obj):
+                # dynamic class parsing will be handled separately
+                pass
+            else:
```

Comment:
Just sanity check: Is `else` all about global variables?

Reply:
yes!